### PR TITLE
(2049) Feature: Delivery Partner creates an adjustment via a form

### PR DESCRIPTION
**Pull request by edavey**
## User story

- So that BEIS can generate accurate reports for SID
- As a DP
- I want to adjust "actual spend" in historic periods in
  line with corrections agreed with BEIS

Trello: https://trello.com/c/Ffp6B5aB/2049-dp-posts-an-adjustment-via-form

## FINANCIAL PERIOD

It's not clear exactly how adjustments should be located in
time, so for now an adjustment:

- is automatically associated with a report in the "active" state, and
- must be assigned to an explicit financial period

## COMMENTS

It's agreed that it's essential that an adjustment includes
an explanation (or narrative) describing the background for
and rationale for the adjustment.

This is done with the new FlexibleComment class which is
going to replace the existing Comment class:

- `Refund` (done)
- `Adjustment` (here)
- `Activity` (to follow, "soon")

## ADJUSTMENT TYPE

Adjustments must also be labelled as applying to either:

1. an Actual spend, or
2. a Refund

When reviewing spending BEIS use the "Download spending
breakdown as CSV file" link on a report page. This uses
`ActivitySpendingBreakdown` to output a CSV.
The downloaded file includes 3 columns per financial period
which concernAdjustment and Refund. To
represent spending, taking proper account of adjustments and
refunds, these columns should be calculated as follows:

i) *actual spend*: this should EXCLUDE refunds and any
   adjustments to refunds (but INCLUDE adjustments to actual
   spend)

ii) *actual refund:* this should INCLUDE any adjustments to
    refunds

iii) *actual net*: this is the total of actual spend LESS
     refunds. It should INCLUDE adjustments to both refunds
     and actuals

## PRESENTATION

Similar to the approach taken with Refunds, we show
Adjustments in their own section of the Financials tab.

Each Adjustment listed on the Activity Financials tab links
to a 'show' view which provides full detail on the
individual Adjustment, including when it was posted, by
whom, the type of adjustment and the important narrative
comment. We also provide links to the associated Report and
Activity.

### Form, with error messages:

adjustments_no_report_selection

### Adjustment section on Activity (Financials tab):

list_of_adjustments

### An individual adjustment in detail:

adjustment_details

### What are adjustments?

Adjustments are financial entries which are used to correct
mistakes. They are used _after_ a report has been approved;
before that point entries (actuals, refunds) can be "changed"
or edited. Design options which were considered:

#### 1) Replicate transaction

Make a new table which is very similar to Transaction. This
was discounted as:

- if an actual is a transaction then it follows that an adjustment
  is a transaction

- duplicating the common fields (`value`, `activity_id`, `report_id`,
  `financial_year`, `financial_quarter`) is likely to have an
  ongoing maintenance cost as our understanding of RODA finance develops

#### 2) Sub-class transaction

Use single-table inheritance to say that Adjustment "is a transaction"
`Adjustment < Transaction`. This feels like a correct approach,
and is attractive. It would require a significant change to the
existing Transaction class and involve a data migration, e.g.
to set the `Transaction#type` on existing records to 'actual'
as distinct from `'adjustment'`

This is the approach taken here, and follows on from the earlier work of:

- modelling actual spend with the `Actual` class as a subclass of `Transaction` using single table inheritance on the `transactions` table
- refactoring Refund to also using STI on Transaction

#### 3) Extend transaction

~~The approach **proposed in this PR** is to introduce a small
`Adjustment` entity to model the concept and meta-data (just
`user_id` and `comment`) for this type of correction.
Adjustment delegates as much as possible to Transaction.~~

**Following discussion and agreement to use approach 3 of STI this PR has been reworked**

## Next steps

- [ ] Is an ADR required? An ADR should be added if this PR introduces a change to the architecture.
- [ ] Is a changelog entry required? An entry should always be made in `CHANGELOG.md`, unless this PR is a small tweak which has no impact outside the development team.
- [ ] Do any environment variables need amending or adding?
- [ ] Have any changes to the XML been checked with the IATI validator? See [XML Validation](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/develop/doc/xml-validation.md)

### CODE REVIEW

File: spec/features/staff/users_can_create_an_adjustment_spec.rb
Comment (pezholio) [inline spec/features/staff/users_can_create_an_adjustment_spec.rb:18]: I like the Cucumber-esque approach here 👍

**Review by pezholio — APPROVED** [review]
Just a couple of tiny things, but overall I like the approach 👍

File: spec/models/adjustment_spec.rb
Comment (edavey) [reply spec/models/adjustment_spec.rb:25]: hmm, interesting option, but no -- I think the job here is to describe the actual interface on `Adjustment` as clearly as possible.

What I'm try to say is "when you send the `create` message to `Adjustment` with an `activity` param, that activity is actually assigned to a `Transaction` object"...

File: spec/policies/activity_policy_spec.rb
Comment (edavey) [reply spec/policies/activity_policy_spec.rb:191]: yes we could do that -- and probably it will be clearer. I saw that we already had a spec which happens to put the report in the necessary 'approved' state and rather than build more objects I tried to:

- use what I had to hand and avoid more setup
- make it clear on line 190 that the Report is in the approved state

but I think it's fair to say that the result is actually more confusing!

I will take another run at this...

File: app/controllers/staff/adjustments_controller.rb
Comment (edavey) [reply app/controllers/staff/adjustments_controller.rb:19]: maybe, I'm a bit sketchy on Pundit. I'll take another look at this...

File: app/models/adjustment.rb
Comment (edavey) [reply app/models/adjustment.rb:6]: good question. I *think* that where we're headed is a 'base' Transaction model which is used by various financial classes:

- `Actual`
- `Adjustment`
- `Refund`

i.e. that transaction is a technical term which can be limited to the model layer so that the user interface, controllers, and service objects, importers, exporters etc can all think in terms of `Actual`, `Refund` and `Adjustment`. The fact that the common financial attributes are delegated to `Transaction` are implementation details at the model layer.

File: spec/policies/activity_policy_spec.rb
Comment (edavey) [reply spec/policies/activity_policy_spec.rb]: I've fixed this up. Turns out it was super-easy to add:

```
before do
  report.update(state: :approved)
end
```

thanks for bringing it up.

File: app/controllers/staff/adjustments_controller.rb
Comment (edavey) [reply app/controllers/staff/adjustments_controller.rb]: yes, very good point. Not much point authorising _after_ the Adjustment is created!

I got rather confused with my authorisations, I think because of the "form object". A key learning has been the syntax for telling pundit which policy to use for an arbitrary object:

```
authorize(@crab_apple, policy_class: ApplePolicy)
```

**Comment by edavey** [comment]
@pezholio I think I've addressed all your points. Thanks.

I won't merge this for now as I know @mec is planning to have a look...

**Comment by mec** [comment]
This is a super interesting PR for me on so many different levels!

I am not going to dig into the code as you have a review already and I think any comments I have will be at a conceptual level.

Top down I am wondering what the user might think about what the expectation is for adding one of these adjustments. As far as I undersatand it, this would be a new thing for them to be doing as in the old spreadsheet driven service, users would provide a single value that could incoporate the three conecpts of any `actual spend`, `refunds` or `adjustments` providing a single figure that ultimately provides BEIS with a 'balanced book'.

Now we are asking for more details on what the adjustment is doing, we are asking users to document the adjustment (which is a good thing!). So what are they doing?

First, I think it is fair to say that the adjustment is about `actual spend` only. This is really the only financial data we collect that might be considered 'fact' (budget and forecast are closer to estimates and other funding which is not BEIS funding so doesn't count towards the idea of a 'balanced BEIS book').

Actual spend is sliced per financial quarter and could be thought of as a total for that quarter:

> I actually spent £ 100,000.00 in qurter two 2021-2022

or as a collection of acual spend 'entries' for the fianncial quarter that add up to that total:

> I actually spend £ 20,000.00 in quarter two 2021-2022 at Tesco
> I acually spent £ 50,000.00 in quarter two 2021-2022 at Morissons
> I actually spent £ 30,000.00 in quarter two 2021-2022 at Sainsburys
> I spent a total of £100,000.00 in quarter two 2021-2022

HIstorically, in the spreadsheets, users were only asked to provide that single total value, and that behaviour remains and is acceptatble to BEIS in the short to mid-term as the service matures and users are more able to provide further breakdown of spend, we were asked to make the application support both.

So we know that an `adjustment` here can be about that total and that it is acceptatble in either case, it is less important to know which supermarket we are adjusting the spend at than it is to get the adjustment correct and the books balanced, in the example above (I am not saying that knowing which supermarket is not valuable, just that right now it is less so).

I like words to describe these (user needs?!) so to summerise where I think we are now:

> I want to adjust the total actual spend value I gave BEIS for financial quarter two 2021-2022 as I have since learnt that it was supplied incorrectly.

Let's talk about `report`. A report is a RODA concept we introduced to control when a delivery partner can make changes to the data set and an attempt at a history of sorts: 

> This is what was reported in financial quarter two 2021-2022.

It's for this reason that I would propose that a user adding an adjustment is more likely to know and understand the financial quarter they are adjusting than the report that aligns to that same financial quater. Right now we have symmetry, one report per financial quarter, but this may not always be the case, the system is designed to accpet multiple reports per financial quarter.

Phew! Sorry that is a lot of information, I wanted to backup my comments with my rational and thought process, it also helps we get everything straight in my mind!

I would propose that an adjustment is:

- a value, amount in GBP
- either negative or positive to account for under or over reported spend
- a financial quarter that identifies the total being adjusted
- the narrative (comment) on why the adjustment is required
- the value cannot be zero as that would not adjust anything
- the financial quarter must be in the past as you cannot adjust spend that has not happened yet
- the report in which the adjustment was added, this is automatic and not something we would ask the user for

And that here, the user selecting a report that must be approved, is going to be a difficult thing for a user to do and I am not sure of the value to BEIS in asking or stroing that.

We can therefore agree and set a clear expectation on what the financial quarter we are asking for is and what the expectation is, I would start with something like:

> Provide the finnacial quarter in which the actual spend took place you wish to adjust

Ultimately, even if that spend took place many financial quarters ago the finances would balanace out with just the adjustment, this is what we think already happens in the spreadsheet, but here we can be more specific.

I will add this comment now and continue to talk abou the modelling/imlpementation as I don't want to loose all this typing!

**Comment by mec** [comment]
I think @CristinaRO comment speaks to why we have `transaction` in the first place, we used transaction as a generic term following IATI, there transactions have many different types and represent many different movements of money (IATI docs if interested: https://iatistandard.org/en/iati-standard/203/activity-standard/iati-activities/iati-activity/transaction/).

Since using transaction as a generic term, we have learnt that the historic data stored in out tables could be a combination of what we now think of as either actual spend, refund or adjustment or a combination of all three and that there is no way for us to know which we have in the historic data, only BEIS and the right devliery partners could untangle that or even decide on the value of doing so.

We also know that (at least in IATI terms) BEIS consider all of these transactions `disbursemnts` (see https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/3123b5ff6d8da3bd5108d6a048e9be09885e9fe9/app/services/create_transaction.rb#L17). And that this concept of disbursement in BEIS language is 'actual spend'.

We looked to add the concepts refunds and adjustments to the roster of transactions the applcation accepts. And this is where we get to implementation details.

As beautifully demostrated by @edavey right here, we have three options and as ever, each have their strengths and weaknesses. I think what is important is that we model these three objects the same way, so it is time to make that decision (my understanding is that otherwise we are 1. Building in debt with the new one 2. Not dealling with exsiting debt) so which route do we take?

Weighing them up very loosely:

- multiple models (as refund) is nice and simple but leads to duplication and querying multiple tables 
- extend the model could be seen to be less approachable and still has multiple tables
- single model, approachable and a single table

As Ed has already said, the single model/table (single table inheritance) feels like the right choice but will involve the most work. That said we have already identified that we want to pay off the debt of having a `transaction` model that models `actual spend` and that taking on this work would resolve that (potentially giving us a transaction model with the type `actual spend` for what BEIS think of as 'Actual spend', etc).

I also believe now is the time to do this as there is no data collected from users for either refunds or adjustments, which makes me far more confident about making changes.

We would still need to look at how doing this might effect other priorities ahead of Q2 reporting and carefully manage expectations of what we are doing.

To summerise:

- single table inheritance is the route I think we should take
- we already have debt we would like to clear around what `transaction` models and this would be paid off as part of this work

File: app/models/adjustment.rb
Comment (CristinaRO) [inline app/models/adjustment.rb:23]: Is there a way to automatically delegate the setters, too? I feel like we're losing some of the benefits of delegation if we have to separately remember to define the setters.

File: app/models/transaction.rb
Comment (CristinaRO) [inline app/models/transaction.rb:23]: This way of retrieving "actual" transactions by defining them against another entity seems a bit brittle and prone to errors. 😬  Is this an intermediate stage before going to another paradigm?

It's making "actuals" second class citizens, in a way.

**Comment by edavey** [comment]
@mec this has been reworked to:

- use single table inheritance, as is already now done by `Actual` and `Refund`
- incorporate the need identified yesterday to capture an "adjustment type" or either "Actual" or "Refund"

I haven't yet changed the behaviour around associating a report. At present an approved report must be selected when completing the form.

Do you want me to make that change here before merging and offering up for user testing? The change -- as I understand it -- would be that we'd automatically assign an approved report to the adjustment. As there can be more than one report for a financial period, we'd need to choose (say) the first or the last...

**Review by mec — APPROVED** [review]
I am in two minds about your question around the report association. As we discovered yesterday, there are concerns about what a report actually is and what it tries to accomplish that we need to get to the bottom of.

My view is still that a user would not be asked to supply a report for any object, as that association is simply 'the report in which you added the change', so always the 'current' report. Reports exist to provide some information on what changed when, but you could argue that a more complete log might meet that need better. - we should meet and discuss as soon as we can.

The way I see `adjusment` is that the financial quarter would be how the user and the service identifies which value to apply the adjustment to, something like: 'the total spend for Q1 2021-2022 is being adjusted by +10000' and then 'and we were alerted to that 'mistake' in the Q2 2021-2022 report'.

If we merge this as is, we risk collecting data where the user has decided what value to supply for report that may or may not turn out to be detrimental later should we change course.

I do think that being able to collect the adjustment sepertately from the actual spend (or refund now!) is where the real value lies so on balance, I am hitting approve now, on the understanding that we meet and go over the concept of report and decide on a course of action and that, as a completely new item, we can migrate all adjustemnts colected after Q2 reports the association we decide is correct.

Apologies for the bit comment - happy to talk this through as needed, just let me know.

🚢

**Comment by edavey** [comment]
@mec this is now revised so that the associated report is set automatically and the DP posting the adjustment does not have to make this selection.

**Comment by mec** [comment]
@edavey I'm really sorry that I am struggling to communicate on this one! I think it would be useful to talk this over in a wider conversation about reports?

I would still expect the `Adjustment` to be assoiciated with the current `active` report, not an approved one, there can only ever be one active report per fund, so we know exactly the one to associate.

We are adjusting something that has already happend, so it will have happend in a past report, therefore that report would be approved, but I don't think that is what is valuable about it.

What I believe is valuable is that in a given reporting period a DP has added an adjustment, we want to surface that so a BEIS user can know that it has happened and deal with any issues that might arise from that adjustment.

The financial quarter and year still pin point exactly which total the adjustment is altering and now we have the `adjustment_type` to tell us if it is adjusting actual spend or a refund.

Associating to the 'current' `active` reports means we can surface all the adjustments that a DP has reported in a 'report' to users as we do for other entities on the report page.

It should also mean that if that `Report` is still `active` the DP user can delete that adjustment entry as the report has not be assured by BEIS at that point, this is also a similar pattern we have.

I am aware how many times I have type 'report' here! As I said, I think there is value in talking this over, let me know?

**Comment by edavey** [comment]
ok, @mec I think I get it now! Sorry for being so slow on the uptake!

**Comment by mec** [comment]
@edavey it's absolutely fine! I am begining to question it all, so that chat about report is still valid and value from my perspective!

**Comment by edavey** [comment]
@mec are we happy to merge this now?

**Review by mec — APPROVED** [review]
🚢

**Comment by mec** [comment]
@edavey yes! Let's ship this! 😆

**Pull request merged**
